### PR TITLE
World Hooks: Isolate apps launched from Kano World

### PR DESCRIPTION
**Pull request by tombettany**
https://trello.com/c/iD5hz3Kf/1028-p1-sonic-pi---system-crash-open-kano-world--launch-sonic-pi-share--play-around-on-sonic-pi--click-share--system-crashes

When an app is launched via the Kano XDG scheme from the browser, the
browser is the parent of the process and hence all child processes. This
poses a problem in Sonic Pi when the share window is launched and causes
the window to dump its core, however this is only a problem if the
browser is the parent of the process. As such, migrate all browser-based
launches to be run through SystemD.

cc @skarbat @radujipa

### CODE REVIEW

**Comment by tombettany** [comment]
[(P1) Sonic Pi - System crash. Open Kano World > Launch Sonic Pi Share > play around on Sonic Pi > Click share > system crashes](https://trello.com/c/iD5hz3Kf/1028-p1-sonic-pi---system-crash-open-kano-world--launch-sonic-pi-share--play-around-on-sonic-pi--click-share--system-crashes)

File: bin/kano-world-hook [hunk 1]
```diff
@@ -28,46 +36,56 @@ from kano.logging import logger
...
-    logger.error(msg)
-    sys.exit(0)
+    # First element is the scheme specifier - should be `kano`
+    try:
+        scheme = url_data.pop(0)
+        assert(scheme=='kano')
+    except:
+        msg = "Could not recognise the {} scheme".format(scheme)
+        logger.error(msg)
+        sys.exit(0)
```
Comment (radujipa) [inline bin/kano-world-hook]: Do we want to exit with `0` for all these cases in here?

File: bin/kano-world-hook [hunk 2]
```diff
@@ -28,46 +36,56 @@ from kano.logging import logger
...
 
-if hasattr(hook_module, "launch"):
-    hook_module.launch(result)
 
-sys.exit(0)
+if __name__ == '__main__':
+    ARGS = docopt.docopt(__doc__)
+    SCHEME_URL = ARGS.get('<kano-scheme-url>')
+
+    main(SCHEME_URL, **ARGS)
```
Comment (radujipa) [inline bin/kano-world-hook]: Not really important, but another way of doing it is to:
```
sys.exit(main(SCHEME_URL, **ARGS))
```
and use `return RC_x` inside the `main` function.

File: kano_world/hooks/share.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # share.py
 #
-# Copyright (C) 2014-2015 Kano Computing Ltd.
+# Copyright (C) 2014-2015, 2017 Kano Computing Ltd.
```
Comment (radujipa) [inline kano_world/hooks/share.py]: Perhaps all these could simply be `2014 - 2017`

**Comment by tombettany** [comment]
Thanks @radujipa. I've made the suggested changes. Merging now.

File: bin/kano-world-hook [hunk 4]
```diff
@@ -26,48 +34,73 @@ if __name__ == '__main__' and __package__ is None:
...
 # TODO: Put kano-dialogs where appropriate
 
-if len(sys.argv) < 2:
-    msg = "{} run with no arguments".format(sys.argv[0])
-    logger.error(msg)
-    sys.exit(0)
+def main(scheme_url, **args):
+    # Parse the URL
+    try:
+        url_data = scheme_url.split(":")
```
Comment (skarbat) [inline bin/kano-world-hook]: Just for future reference, you might prefer to use `urlparse`here.

 * https://docs.python.org/2/library/urlparse.html

Comment (tombettany) [inline bin/kano-world-hook]: Nice find Albert. I didn't want to make too drastic a change for this fix but we should definitely migrate to use that after release.

**Pull request merged**